File: Src/java/engine/src/test/java/org/opencds/cqf/cql/engine/execution/BacktraceTest.java
package org.opencds.cqf.cql.engine.execution;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;
import static org.junit.jupiter.api.Assertions.fail;
import static org.opencds.cqf.cql.engine.execution.BacktraceTest.ContextMatcher.hasContext;
import static org.opencds.cqf.cql.engine.execution.BacktraceTest.ExpressionDefinitionMatcher.isExpressionDefinition;
import static org.opencds.cqf.cql.engine.execution.BacktraceTest.FunctionMatcher.isFunction;
import static org.opencds.cqf.cql.engine.execution.BacktraceTest.VariableMatcher.isVariable;

import java.util.Objects;
import org.hamcrest.Description;
import org.hamcrest.TypeSafeMatcher;
import org.hl7.elm.r1.ExpressionDef;
import org.hl7.elm.r1.FunctionDef;
import org.junit.jupiter.api.Test;
import org.opencds.cqf.cql.engine.exception.Backtrace;
import org.opencds.cqf.cql.engine.exception.CqlException;

class BacktraceTest extends CqlTestBase {

    static class FunctionMatcher extends TypeSafeMatcher<FunctionDef> {

        private final String expectedName;

        public FunctionMatcher(final String expectedName) {
            this.expectedName = expectedName;
        }

        @Override
        protected boolean matchesSafely(final FunctionDef functionDefinition) {
            return Objects.equals(functionDefinition.getName(), expectedName);
        }

        @Override
        public void describeTo(final Description description) {
            description.appendText("a function named ").appendValue(expectedName);
        }

        public static FunctionMatcher isFunction(final String expectedName) {
            return new FunctionMatcher(expectedName);
        }
    }

    static class ExpressionDefinitionMatcher extends TypeSafeMatcher<ExpressionDef> {

        private final String expectedName;

        public ExpressionDefinitionMatcher(final String expectedName) {
            this.expectedName = expectedName;
        }

        @Override
        protected boolean matchesSafely(final ExpressionDef expressionDefinition) {
            return Objects.equals(expressionDefinition.getName(), expectedName);
        }

        @Override
        public void describeTo(final Description description) {
            description.appendText("an expression definition named ").appendValue(expectedName);
        }

        public static ExpressionDefinitionMatcher isExpressionDefinition(final String expectedName) {
            return new ExpressionDefinitionMatcher(expectedName);
        }
    }

    static class VariableMatcher extends TypeSafeMatcher<Variable> {

        private final String expectedName;
        private final Object expectedValue;

        public VariableMatcher(final String expectedName, final Object expectedValue) {
            this.expectedName = expectedName;
            this.expectedValue = expectedValue;
        }

        @Override
        public boolean matchesSafely(final Variable variable) {
            return Objects.equals(variable.getName(), expectedName)
                    && Objects.equals(variable.getValue(), expectedValue);
        }

        @Override
        public void describeTo(final Description description) {
            description
                    .appendText("an argument with name ")
                    .appendValue(expectedName)
                    .appendText(" and value ")
                    .appendValue(expectedValue);
        }

        public static VariableMatcher isVariable(final String expectedName, final Object expectedValue) {
            return new VariableMatcher(expectedName, expectedValue);
        }
    }

    static class ContextMatcher extends TypeSafeMatcher<Backtrace.FunctionoidFrame> {

        private final String expectedName;
        private final Object expectedValue;

        public ContextMatcher(final String expectedName, final Object expectedValue) {
            this.expectedName = expectedName;
            this.expectedValue = expectedValue;
        }

        @Override
        public boolean matchesSafely(Backtrace.FunctionoidFrame frame) {
            return Objects.equals(frame.getContextName(), expectedName)
                    && Objects.equals(frame.getContextValue(), expectedValue);
        }

        @Override
        public void describeTo(Description description) {
            description
                    .appendText("a context with name ")
                    .appendValue(expectedName)
                    .appendText(" and value ")
                    .appendValue(expectedValue);
        }

        public static ContextMatcher hasContext(final String expectedName, final Object expectedValue) {
            return new ContextMatcher(expectedName, expectedValue);
        }
    }

    @Test
    public void backtrace_smoke() {
        try {
            engine.evaluate("BacktraceTest");
            fail("Expected exception but none was thrown");
        } catch (CqlException e) {
            assertThat(e.getBacktrace(), notNullValue());
            final var backtrace = e.getBacktrace();
            final var frames = backtrace.getFrames();
            assertThat(frames, hasSize(3));
            // Frame 1
            final var frame1 = frames.get(0);
            assertThat(frame1, instanceOf(Backtrace.FunctionoidFrame.class));
            final var frame1Functionoid = (Backtrace.FunctionoidFrame) frame1;
            final var definition1 = frame1Functionoid.getDefinition();
            assertThat(definition1, instanceOf(FunctionDef.class));
            assertThat((FunctionDef) definition1, isFunction("F"));
            assertThat(frame1Functionoid.getArguments(), contains(isVariable("X", 3)));
            assertThat(
                    frame1Functionoid.getLocalVariables(), containsInAnyOrder(isVariable("A", 7), isVariable("_", 1)));
            assertThat(frame1Functionoid, hasContext("Unfiltered", null));
            // Frame 2
            final var frame2 = frames.get(1);
            assertThat(frame2, instanceOf(Backtrace.FunctionoidFrame.class));
            final var frame2Functionoid = (Backtrace.FunctionoidFrame) frame2;
            final var definition2 = frame2Functionoid.getDefinition();
            assertThat(definition2, instanceOf(FunctionDef.class));
            assertThat((FunctionDef) definition2, isFunction("G"));
            assertThat(frame2Functionoid.getArguments(), contains(isVariable("Y", 2), isVariable("Z", "hi")));
            assertThat(frame2Functionoid.getLocalVariables(), empty());
            assertThat(frame2Functionoid, hasContext("Unfiltered", null));
            assertThat(frame2Functionoid.getContextValue(), equalTo(null));
            // Frame 3
            final var frame3 = frames.get(2);
            assertThat(frame3, instanceOf(Backtrace.FunctionoidFrame.class));
            final var frame3Functionoid = (Backtrace.FunctionoidFrame) frame3;
            final var definition3 = frame3Functionoid.getDefinition();
            assertThat(definition3, instanceOf(ExpressionDef.class));
            assertThat(definition3, isExpressionDefinition("E2"));
            assertThat(frame3Functionoid.getArguments(), empty());
            assertThat(frame3Functionoid.getLocalVariables(), empty());
            assertThat(frame3Functionoid, hasContext("Unfiltered", null));
        }
    }
}

File: Src/java/engine/src/test/kotlin/org/opencds/cqf/cql/engine/execution/BacktraceTest.kt
package org.opencds.cqf.cql.engine.execution

import org.hamcrest.Description
import org.hamcrest.MatcherAssert
import org.hamcrest.Matchers
import org.hamcrest.TypeSafeMatcher
import org.hl7.elm.r1.ExpressionDef
import org.hl7.elm.r1.FunctionDef
import org.hl7.elm.r1.VersionedIdentifier
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.opencds.cqf.cql.engine.exception.Backtrace.FunctionoidFrame
import org.opencds.cqf.cql.engine.exception.CqlException

internal class BacktraceTest : CqlTestBase() {
    internal class FunctionMatcher(private val expectedName: String?) :
        TypeSafeMatcher<FunctionDef>() {
        override fun matchesSafely(functionDefinition: FunctionDef): Boolean {
            return functionDefinition.name == expectedName
        }

        override fun describeTo(description: Description) {
            description.appendText("a function named ").appendValue(expectedName)
        }

        companion object {
            fun isFunction(expectedName: String?): FunctionMatcher {
                return FunctionMatcher(expectedName)
            }
        }
    }

    internal class ExpressionDefinitionMatcher(private val expectedName: String?) :
        TypeSafeMatcher<ExpressionDef>() {
        override fun matchesSafely(expressionDefinition: ExpressionDef): Boolean {
            return expressionDefinition.name == expectedName
        }

        override fun describeTo(description: Description) {
            description.appendText("an expression definition named ").appendValue(expectedName)
        }

        companion object {
            fun isExpressionDefinition(expectedName: String?): ExpressionDefinitionMatcher {
                return ExpressionDefinitionMatcher(expectedName)
            }
        }
    }

    internal class VariableMatcher(
        private val expectedName: String?,
        private val expectedValue: Any?,
    ) : TypeSafeMatcher<Variable>() {
        override fun matchesSafely(variable: Variable): Boolean {
            return variable.name == expectedName && variable.value == expectedValue
        }

        override fun describeTo(description: Description) {
            description
                .appendText("an argument with name ")
                .appendValue(expectedName)
                .appendText(" and value ")
                .appendValue(expectedValue)
        }

        companion object {
            fun isVariable(expectedName: String?, expectedValue: Any?): VariableMatcher {
                return VariableMatcher(expectedName, expectedValue)
            }
        }
    }

    internal class VersionedIdentifierMatcher(
        private val expectedId: String?,
        private val expectedVersion: String?,
    ) : TypeSafeMatcher<VersionedIdentifier>() {
        override fun matchesSafely(vi: VersionedIdentifier): Boolean {
            return true
        }

        override fun describeTo(description: Description) {
            description
                .appendText("a library id")
                .appendValue(expectedId)
                .appendText("a library version")
                .appendValue(expectedVersion)
        }

        companion object {
            fun isVersionedIdentifier(
                expectedId: String?,
                expectedVersion: String?,
            ): VersionedIdentifierMatcher {
                return VersionedIdentifierMatcher(expectedId, expectedVersion)
            }
        }
    }

    internal class ContextMatcher(
        private val expectedName: String?,
        private val expectedValue: Any?,
    ) : TypeSafeMatcher<FunctionoidFrame>() {
        override fun matchesSafely(frame: FunctionoidFrame): Boolean {
            return frame.contextName == expectedName && frame.contextValue == expectedValue
        }

        override fun describeTo(description: Description) {
            description
                .appendText("a context with name ")
                .appendValue(expectedName)
                .appendText(" and value ")
                .appendValue(expectedValue)
        }

        companion object {
            fun hasContext(expectedName: String?, expectedValue: Any?): ContextMatcher {
                return ContextMatcher(expectedName, expectedValue)
            }
        }
    }

    @Test
    fun backtrace_smoke() {
        try {
            engine.evaluate { library("BacktraceTest") }.onlyResultOrThrow
            Assertions.fail<Any?>("Expected exception but none was thrown")
        } catch (e: CqlException) {
            MatcherAssert.assertThat(e.backtrace, Matchers.notNullValue())
            val backtrace = e.backtrace
            val frames = backtrace.frames
            MatcherAssert.assertThat(frames, Matchers.hasSize(3))
            // Frame 1
            val frame1 = frames!![0]
            MatcherAssert.assertThat(frame1, Matchers.instanceOf(FunctionoidFrame::class.java))
            val frame1Functionoid = frame1 as FunctionoidFrame
            val definition1 = frame1Functionoid.definition
            MatcherAssert.assertThat(definition1, Matchers.instanceOf(FunctionDef::class.java))
            MatcherAssert.assertThat<FunctionDef?>(
                definition1 as FunctionDef?,
                FunctionMatcher.isFunction("F"),
            )
            MatcherAssert.assertThat(
                frame1Functionoid.arguments,
                Matchers.contains(VariableMatcher.isVariable("X", 3)),
            )
            MatcherAssert.assertThat(
                frame1Functionoid.localVariables,
                Matchers.containsInAnyOrder(
                    VariableMatcher.isVariable("A", 7),
                    VariableMatcher.isVariable("_", 1),
                ),
            )
            MatcherAssert.assertThat(
                frame1Functionoid.libraryIdentifier,
                VersionedIdentifierMatcher.isVersionedIdentifier(
                    "BacktraceTest",
                    expectedVersion = null,
                ),
            )
            MatcherAssert.assertThat(
                frame1Functionoid,
                ContextMatcher.hasContext("Unfiltered", null),
            )
            // Frame 2
            val frame2 = frames[1]
            MatcherAssert.assertThat(frame2, Matchers.instanceOf(FunctionoidFrame::class.java))
            val frame2Functionoid = frame2 as FunctionoidFrame
            val definition2 = frame2Functionoid.definition
            MatcherAssert.assertThat(definition2, Matchers.instanceOf(FunctionDef::class.java))
            MatcherAssert.assertThat<FunctionDef?>(
                definition2 as FunctionDef?,
                FunctionMatcher.isFunction("G"),
            )
            MatcherAssert.assertThat(
                frame2Functionoid.arguments,
                Matchers.contains(
                    VariableMatcher.isVariable("Y", 2),
                    VariableMatcher.isVariable("Z", "hi"),
                ),
            )
            MatcherAssert.assertThat(frame2Functionoid.localVariables, Matchers.empty())
            MatcherAssert.assertThat(
                frame2Functionoid.libraryIdentifier,
                VersionedIdentifierMatcher.isVersionedIdentifier(
                    "BacktraceTest",
                    expectedVersion = null,
                ),
            )
            MatcherAssert.assertThat(
                frame2Functionoid,
                ContextMatcher.hasContext("Unfiltered", null),
            )
            MatcherAssert.assertThat(frame2Functionoid.contextValue, Matchers.equalTo<Any?>(null))
            // Frame 3
            val frame3 = frames[2]
            MatcherAssert.assertThat(frame3, Matchers.instanceOf(FunctionoidFrame::class.java))
            val frame3Functionoid = frame3 as FunctionoidFrame
            val definition3 = frame3Functionoid.definition
            MatcherAssert.assertThat(definition3, Matchers.instanceOf(ExpressionDef::class.java))
            MatcherAssert.assertThat(
                definition3,
                ExpressionDefinitionMatcher.isExpressionDefinition("E2"),
            )
            MatcherAssert.assertThat(frame3Functionoid.arguments, Matchers.empty())
            MatcherAssert.assertThat(frame3Functionoid.localVariables, Matchers.empty())
            MatcherAssert.assertThat(
                frame3Functionoid.libraryIdentifier,
                VersionedIdentifierMatcher.isVersionedIdentifier(
                    "BacktraceTest",
                    expectedVersion = null,
                ),
            )
            MatcherAssert.assertThat(
                frame3Functionoid,
                ContextMatcher.hasContext("Unfiltered", null),
            )
        }
    }
}
